From 2349f32ee3c8934a9a5aee2c72ccb1b954f18d6e Mon Sep 17 00:00:00 2001
From: Robin Humble <plaguedbypenguins@gmail.com>
Date: Mon, 2 Sep 2013 23:02:39 +1000
Subject: [PATCH] workaround an armv6 gcc 4.7 bug

this fixes com.android.phone getting a SIGBUS on startup due to mmap'd
tzdata reads that always crash in this libcore code.
gcc 4.8 works without issue and 4.6 worked fine with cm10.1, so this
workaround has been made specific to just gcc 4.7.

Change-Id: Ide3b33dd8bdbb1e0f1a91b65ba21c7aa8e276af5

diff --git a/luni/src/main/native/libcore_io_Memory.cpp b/luni/src/main/native/libcore_io_Memory.cpp
index 77aef5b..8a491c6 100644
--- a/luni/src/main/native/libcore_io_Memory.cpp
+++ b/luni/src/main/native/libcore_io_Memory.cpp
@@ -110,7 +110,11 @@ static inline void swapInts(jint* dstInts, const jint* srcInts, size_t count) {
     if ((reinterpret_cast<uintptr_t>(dstInts) & INT_ALIGNMENT_MASK) == 0 &&
         (reinterpret_cast<uintptr_t>(srcInts) & INT_ALIGNMENT_MASK) == 0) {
         for (size_t i = 0; i < count; ++i) {
+#if (__GNUC__ == 4) && (__GNUC_MINOR__ == 7) && defined(__ARM_ARCH_5TE__)   // work around gcc 4.7 bug on armv6
+            jint v = ((jint *)srcInts)[i];
+#else
             jint v = *srcInts++;
+#endif
             *dstInts++ = bswap_32(v);
         }
     } else {
